src/ServiceControl/Recoverability/Retrying/RetryProcessor.cs

```diff
@@ -195,6 +198,57 @@ async Task<int> Stage(RetryBatch stagingBatch, IAsyncDocumentSession session)
             return messages.Length;
         }
 
+        async Task TryStageMessage(FailedMessage message, string stagingId, FailedMessageRetry failedMessageRetry)
+        {
+            try
+            {
+                await StageMessage(message, stagingId).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                if (failedMessageRetry.StageAttempts < 5)
+                {
+                    await IncrementAttemptCounter(failedMessageRetry).ConfigureAwait(false);
+                    //We need to rethrow here to stop the current staging process
+                    throw new Exception($"Error while trying to stage failed message {message.UniqueMessageId}", e);
+                }
+
+                Log.Error($"Failed message {message.UniqueMessageId} reached its staging retry limit ({failedMessageRetry.StageAttempts}) and is going to be removed from the batch.", e);
+
+                await store.AsyncDatabaseCommands.DeleteAsync(FailedMessageRetry.MakeDocumentId(message.UniqueMessageId), null)
```

Comment:
Should the delete be the last statement perhaps? This is an atomic command meaning the document is out of the DB, so what would it mean if the domain events fails for some reason? Is it better to re-try the staging or better to not raise a domain event?

Reply:
Not sure here. If we flip that then we can have a situation where we announced that the document has been removed from the batch while in reality it might succeed in the next staging attempt. Currently the event is only used to created an activity log entry. I don't have any preference here. @mikeminutillo ?